crates/derive/src/online/blob_provider.rs

```diff
@@ -176,6 +176,168 @@ where
+    S: SlotDerivation,
+> {
+    primary: OnlineBlobProvider<B, S>,
+    fallback: F,
+    _slot_derivation: PhantomData<S>,
+}
+
+impl<B: BeaconClient, F: BlobSidecarProvider, S: SlotDerivation>
+    OnlineBlobProviderWithFallback<B, F, S>
+{
+    /// Creates a new instance of the [OnlineBlobProviderWithFallback] with the
+    /// specified primary and fallback providers.
+    pub fn new(primary: OnlineBlobProvider<B, S>, fallback: F) -> Self {
+        Self { primary, fallback, _slot_derivation: PhantomData }
+    }
+
+    /// Attempts to fetch blob sidecars from the fallback provider.
+    async fn fallback_fetch_filtered_sidecars(
+        &self,
+        block_ref: &BlockInfo,
+        blob_hashes: &[IndexedBlobHash],
+    ) -> Result<Vec<BlobSidecar>, BlobProviderError> {
+        if blob_hashes.is_empty() {
+            return Ok(Vec::new());
+        }
+
+        // Extract the genesis timestamp and slot interval from the primary provider.
+        let slot = S::slot(
+            self.primary.genesis_time.expect("Genesis Config Loaded"),
+            self.primary.slot_interval.expect("Config Spec Loaded"),
+            block_ref.timestamp,
+        )
+        .map_err(BlobProviderError::Slot)?;
+
+        // Fetch blob sidecars for the given block reference and blob hashes.
+        let sidecars = self.fallback.beacon_blob_side_cars(slot, blob_hashes).await?;
+
+        // Filter blob sidecars that match the indicies in the specified list.
+        let blob_hash_indicies = blob_hashes.iter().map(|b| b.index).collect::<Vec<_>>();
+        let filtered = sidecars
+            .into_iter()
+            .filter(|s| blob_hash_indicies.contains(&(s.inner.index as usize)))
+            .collect::<Vec<_>>();
+
+        // Validate the correct number of blob sidecars were retrieved.
+        if blob_hashes.len() != filtered.len() {
+            return Err(BlobProviderError::SidecarLengthMismatch(blob_hashes.len(), filtered.len()));
+        }
+
+        Ok(filtered.into_iter().map(|s| s.inner).collect::<Vec<BlobSidecar>>())
+    }
+}
+
+#[async_trait]
+impl<B, F, S> BlobProvider for OnlineBlobProviderWithFallback<B, F, S>
+where
+    B: BeaconClient + Send + Sync,
+    F: BlobSidecarProvider + Send + Sync,
+    S: SlotDerivation + Send + Sync,
+{
+    /// Fetches blob sidecars that were confirmed in the specified L1 block with the given indexed
+    /// hashes. The blobs are validated for their index and hashes using the specified
+    /// [IndexedBlobHash].
+    async fn get_blobs(
+        &mut self,
+        block_ref: &BlockInfo,
+        blob_hashes: &[IndexedBlobHash],
+    ) -> Result<Vec<Blob>, BlobProviderError> {
+        crate::inc!(PROVIDER_CALLS, &["blob_provider", "get_blobs"]);
+        crate::timer!(START, PROVIDER_RESPONSE_TIME, &["blob_provider", "get_blobs"], timer);
+        match self.primary.get_blobs(block_ref, blob_hashes).await {
+            Ok(blobs) => Ok(blobs),
+            Err(_primary_err) => {
```

Comment:
We should log the primary error here. If we're seeing much traffic routed to the fallback provider, we'd want to investigate the logs to see why the primary provider is throwing errors.